docs: life of a request. (#11284)
envoyproxy/envoy: docs/root/intro/life_of_a_request.rst

.. _life_of_a_request:

Life of a Request
=================

Below we follow the events that take place after the arrival of a request at the Envoy proxy from a
downstream client, the dispatch to an upstream backend and the corresponding response path.

Network topology
----------------

Envoy can be used in a wide variety of networking topologies. We focus on the inner operation
of Envoy below, but briefly we address how Envoy relates to the rest of the network in this section.

Envoy originated as a `service mesh
<https://blog.envoyproxy.io/service-mesh-data-plane-vs-control-plane-2774e720f7fc>`_ sidecar proxy,
factoring out load balancing, routing, observability, security and discovery services from
applications. In the service mesh model:

> * Ingress listeners take requests from other nodes in the service mesh and forward them to the

Reviewer: optional, do you think it's worth introducing downstream and upstream here
from "downstream" nodes -> to "upstream" local applications?

It took me months to remember which was "client" and which was "backend" and that was with a little diagram on my desk :-P

Author: Agreed, I've added a terminology section right at the start.

  local application. Responses from the local application flow back through Envoy to the other
  nodes in the service mesh.
* Egress listeners take requests from the local application and forward them to other nodes in the
  network. These nodes will also be typically running Envoy and accepting the request via their
  ingress listeners.

.. note::
  TODO: add a service mesh diagram

Envoy is used in a variety of domains beyond the service mesh. For example, it can also act
as an internal load balancer and as an edge networking ingress/egress proxy.

.. note::
  TODO: add topology diagrams for edge and ILB.

Envoy may be configured in multi-tier topologies for scalability and reliability:

.. note::
  TODO: add topology diagrams for multi-tier.

In all the above cases, a request will arrive at a specific Envoy via TCP, UDP or Unix Domain

> Sockets from a downstream. Envoy will forward requests to upstreams via TCP or Unix Domain

Reviewer: nit: from downstream or from a downstream server

Sockets. We focus on a single Envoy proxy that is part of one of these network topologies
below.

Configuration
-------------

Envoy is a platform with a high degreee of extensibility. This results in a combinatorial
explosion of possible request paths, depending on:

* L3/4 protocol, e.g. TCP, UDP, Unix Domain Sockets.
* L7 protocol, e.g. HTTP/1, HTTP/2, HTTP/3, gRPC, Thrift, Dubbo, Kafka, Redis and various databases.
* Transport socket, e.g. plain text, TLS, ALTS.
* Routing, e.g. PROXY protocol, original destination, dynamic forwarding.
* Authentication and authorization.
* Circuit breakers and outlier detection.
* Many other networking, HTTP, listener, access logging, health checking, tracing and stats
  extensions and configuration options.

It's helpful to limit the nature of the configuration and request to provide an illustrative

> narrative below, so we opt for the following parameters:

Reviewer: This is something I 100% suck at, but it may be worth doing a pass pretending you're ESL (or asking someone who is to do so) and using more common words where you can.

Author: I also am terrible at this. Please correct places that things seem that they aren't ESL friendly. Hopefully we can get a final pass and clean these up at the end.

* An HTTP/2 request with TLS over a TCP connection for both downstream and upstream.
* The HTTP connection manager as the only network filter.

> * The router filter as the only HTTP filter (TODO: should we add more filters?).

Reviewer: I'd suggest doing at least one other if for no other reason than to show that the request passes through RandomFilter then router, and the response goes router -> RandomFilter.

* Filesystem access logging.
* Statsd sink.

We assume a static bootstrap configuration, but this is a control plane detail that is
orthogonal to the data plane:

.. code-block:: yaml

  todo: simple bootstrap example of above

High level architecture
-----------------------

The request processing path in Envoy has responsibilities split across two major loosely
coupled subsystems:

* *ListenerManager* where the downstream client related aspects of request processing take place.
  This includes all processing from when the first bytes are read from the socket until a

>   decoded HTTP request is ready to forward to an upstream cluster. This subsystem is also

Reviewer: upstream cluster

hm, per my point about downstream and upstream, maybe even worth defining some terms at the top?  Or links to definitions? Especially for interns, they may not be familiar with cluster, connection pooling, etc.

Apart from the language pass we could ask someone not proxy savvy to take a read through and tell us what's not obvious, since we've been immersed in this space long enough we're probably going to miss a bunch.  Given the more clear we make this guide, the easier our onboarding is, I think it's worth getting a pass either from a noogler in our org, or some coworkers who aren't doing proxying, so we can catch as many of those before summer as we can.

Author: I agree; I have some folks in mind for this, so will ping them once we have an approximately complete doc.

  responsible for the response path to the client, the downstream HTTP/2 codec lives here.

> * *ClusterManager* where the HTTP request is forwarded to an upstream backend. This is where

Reviewer 2: This might be a bit pedantic but I wouldn't claim that the CM is responsible for forwarding the request, I'd say that the CM is what is responsible for selecting and configuring the upstream connection/stream (including the codec), but that the router is what forwards the request.

  knowledge of cluster and endpoint health, load balancing and connection pooling exists. The
  upstream HTTP/2 codec lives here.

The two subsystems are bridged with the HTTP router filter.

We use the labels *ListenerManager* and *ClusterManager* above to refer to the family of modules and
instance classes that are rooted in these top-level manager classes. In reality, there are numerous
components that we discuss below that are instantiated prior to and during the course of a request
by these management systems, for example listeners, filter chains, codecs, connection pools and load
balancing data structures.

.. note::
  TODO: add architecture diagram.

Request path
------------

Overview
~~~~~~~~

The request path for the initial bytes that constitute the request headers takes the following
steps:

1. A TCP connection from a downstream is accepted by an Envoy listener associated with a worker

>    thread.

Reviewer: maybe worth a server arch paragrah, calling out that each connection will land on a single thread ("worker thread") and stay there for the lifetime of all traffic on that connection.
Threading model is one of the first questions folks tend to ask when picking up Envoy

Author: Yep, added a paragraph under the high-level arch section.

2. The listener filter chain is instantiated and executed, which can provide SNI and other pre-TLS info.
3. On network reads, the TLS transport socket transforms the encrypted TCP connection

>    to a plain text network connection.

Reviewer 2: maybe have this be more about transforming the data read from the connection to plaintext for further processing? this reads like we're getting a new connection

4. The network filter chain is instantiated and executed. The most significant filter for HTTP is
   the HTTP connection manager which sits at the end of the filter chain.

> 5. The HTTP/2 codec in HTTP connection manager lifts the plain text TCP connection to a number of

Reviewer: lifts -> ???
for H2 enthusiastis I'd use demuxes but that's probably not going to clear things up for non-H2 geeks.  transforms?  turns the TCP connection into ...?

>    independent HTTP streams. Each HTTP stream is a single request.

Reviewer: request and response?

6. For each HTTP stream, an HTTP filter chain is instantiated and executed. The most significant
   filter for HTTP is the router filter which sits at the end of the filter chain. Once

>    *decodeHeaders* processing is completed by the HTTP filter chain, the request headers on the stream

Reviewer 2: since the router is part of the HTTP filter chain, forwarding happens during the filter chain, not after

>    are ready to start forwarding to the upstream.

Reviewer 2: maybe include a point about route resolution happening at this point?

7. The router filter requests an HTTP connection pool from the cluster manager for the matched
   cluster.
8. Cluster specific load balancing is performed to find an endpoint. A new connection to the

>    endpoint is created if the endpoint's connection pool is empty.

Reviewer 2: not sure how specific we want to get here: there are cases even with HTTP/2 now that we create more than one connection (via `maxRequestsPerConnection` option).

Author: Updated with a point on capacity. The related section below will be where we can elaborate on these details.

9. The upstream endpoint connection's HTTP/2 codec combines streams from this and other requests into an L4 stream.
10. The upstream endpoint connection's TLS transport socket encrypts these bytes and writes them to
    a TCP socket for the upstream connection.

.. note::
  TODO: high-level request path diagam

Following the request headers, the path setup above processes reads and write events from
the downstream and upstream endpoints on the TCP connection's worker thread. The request
will go through a lifecycle which includes both request and response header processing,
body streaming and trailer processing. Following the completion of a request, post-request
processing will update stats and write to the access log.

> The above description is terse, we elaborate on each of these request processing steps below.

Reviewer: Maybe call that out at the top, so folks know it's an overview and there's detailed explanation below?

.. _life_of_a_request:

Life of a Request
=================

Below we describe the events in the life of a request passing through an Envoy proxy. We first
describe how Envoy fits into the request path for a request and then the internal events that take
place following the arrival of a request at the Envoy proxy from downstream. We follow the request
until the corresponding dispatch upstream and the response path.

Terminology
-----------

Envoy uses the following terms through its codebase and documentation:

* *Cluster*: a logical service with a set of endpoints that Envoy forwards requests to.
* *Downstream*: an entity connecting to Envoy. This may be a local application (in a sidecar model) or
  a network node. In non-sidecar models, this is a remote client.
* *Endpoints*:  nodes that implement a logical service. They are grouped into clusters.
* *Filter*: a module in the request processing pipeline providing some aspect of request handling. An
  analogy from Unix is the composition of small utilities (filters) with Unix pipes (filter chains).
* *Filter chain*: a series of filters.

> * *Listeners*: Envoy module responsible for binding to an IP/port, accepting new TCP connections (or

Author: I'm going to skip this one as it's a bit on the advanced end of the spectrum, but thanks for reminding.

                  port_value: 10002
      http2_protocol_options:
        max_concurrent_streams: 100
  # statsd sink.
  stats_sinks:
     - name: envoy.stat_sinks.statsd
       typed_config:
         "@type": type.googleapis.com/envoy.config.metrics.v2.StatsdSink
         tcp_cluster_name: some_statsd_cluster

High level architecture
-----------------------

The request processing path in Envoy has two main parts:

* :ref:`Listener subsystem <arch_overview_listeners>` which handles **downstream** request
  processing. This includes all processing from when the first bytes are read from the socket until
  the HTTP request is ready to forward to an upstream cluster. It is also responsible for the
  response path to the client, the downstream HTTP/2 codec lives here.
* :ref:`Cluster subsystem <arch_overview_cluster_manager>` which is responsible for selecting and
  configuring the **upstream** connection to an endpoint. This is where knowledge of cluster and
  endpoint health, load balancing and connection pooling exists. The upstream HTTP/2 codec lives
  here.

The two subsystems are bridged with the HTTP router filter, which forwards the HTTP request from
downstream to upstream.

> .. image:: /_static/lor-architecture.svg

Author: It's a good idea, but I haven't introduced the HCM at this point, and that might add some extra detail that doesn't really help clarify the linkage at this stage. If you have any ideas on how to introduce HCM while still being at 10,000 feet, happy to add it.

<arch_overview_threading>` process requests. All threads operate around an event loop (`libevent
<https://libevent.org/>`_) and any given downstream TCP connection (including all the multiplexed
streams on it) will be handled by exactly one worker thread for its lifetime. Each worker thread
maintains its own pool of TCP connections to upstream endpoints. :ref:`UDP
<arch_overview_listeners_udp>` handling makes use of SO_REUSEPORT to have the kernel consistency
hash the source/destination IP:port tuples to the same worker thread. UDP filter state is global to
a worker thread, with the filter responsible for providing session semantics as needed. This is in
contrast to the connection oriented TCP filters we discuss below.

Worker threads rarely share state and operate in a trivially parallel fashion. This threading model
enables scaling to very high core count CPUs.

Request flow
------------

Overview
^^^^^^^^

A brief outline of the life cycle of a request and response using the example configuration above:

1. A TCP connection from downstream is accepted by an Envoy :ref:`listener
   <arch_overview_listeners>` running on a :ref:`worker thread <arch_overview_threading>`.

> 2. The :ref:`listener filter <arch_overview_listener_filters>` chain is created and runs. It can

Reviewer 3: Mmmm I see that you add this detail below. Probably not needed here.

   provide SNI and other pre-TLS info. Once completed, the listener will match a network filter
   chain. Each listener may have multiple filter chains which match on some combination of
   destination IP CIDR range, SNI, ALPN, source ports, etc. A transport socket, in our case the TLS
   transport socket, is associated with this filter chain.
3. On network reads, the :ref:`TLS <arch_overview_ssl>` transport socket decrypts the data read from
   the TCP connection to plain text for further processing.
4. The :ref:`network filter <arch_overview_network_filters>` chain is created and runs. The most
   important filter for HTTP is the HTTP connection manager, which is the last network filter in the
   chain.
5. The HTTP/2 codec in :ref:`HTTP connection manager <arch_overview_http_conn_man>` demultiplexes

>    the plain text TCP connection to a number of independent HTTP streams. Each HTTP stream handles a

Reviewer 4: "plain text TCP connection" is confusing because all TCP connections in the example are TLS. Maybe "plain text data stream", maybe add "from the decrypted TLS connection".

Also, for people not familiar with encryption terminology, it may not be clear that "plain text" doesn't mean text as in HTTP/1 is a text-based protocol and HTTP/2 is a binary protocol. So maybe instead of "plain text data stream" use "decrypted data stream".

   When `decodeHeaders` is invoked on the router filter, the route is selected and a cluster is
   picked. The request headers on the stream are forwarded to an upstream end point in that cluster.
   The :ref:`router <arch_overview_http_routing>` filter requests an HTTP :ref:`connection pool
   <arch_overview_conn_pool>` from the cluster manager for the matched cluster to do this.
7. Cluster specific :ref:`load balancing <arch_overview_load_balancing>` is performed to find an
   endpoint. The cluster’s circuit breakers are checked to determine if a new stream is allowed. A
   new connection to the endpoint is created if the endpoint's connection pool is empty or lacks
   capacity.
8. The upstream endpoint connection's HTTP/2 codec multiplexes the request’s stream with any other
   streams going to that upstream over a single TCP connection.
9. The upstream endpoint connection's TLS transport socket encrypts these bytes and writes them to a
   TCP socket for the upstream connection.
10. The request, consisting of headers, and optional body and trailers, is proxied upstream, and the
    response is proxied downstream. The response passes through the HTTP filters in the opposite

>     order from the request, starting at the router filter and passing through CustomFilter, before

Reviewer 3: link to: arch_overview_http_filters_ordering?

Each :ref:`worker thread <arch_overview_threading>` maintains its own *Listener* instance for each
of the configured listeners. Each listener may bind to the same port via SO_REUSEPORT or share a
single socket bound to this port. When a new TCP connection arrives, the kernel decides which
worker thread will accept the connection and the *Listener* for this worker thread will have its
``Server::ConnectionHandlerImpl::ActiveTcpListener::onAccept()`` callback invoked.

2. Listener filter chains and network filter chain matching
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The worker thread’s *Listener* then creates and runs the :ref:`listener filter
<arch_overview_listener_filters>` chain. Filter chains are created by applying each filter’s *filter
factory*. The factory is aware of the filter’s configuration and creates a new instance of the
filter for each connection or stream.

In the case of our TLS listener configuration, the listener filter chain consists of the :ref:`TLS
inspector <config_listener_filters_tls_inspector>` filter
(``envoy.filters.listener.tls_inspector``). This filter examines the initial TLS handshake and

> extracts the server name (SNI). The SNI is then made available for filter chain matching. The TLS

Reviewer 4: I think we should put it explicitly in the configuration in this example. I think it leads to better understanding of what is going on. And if you need other listener filters (PROXY protocol for instance), they should always be configured explicitly to ensure they're in the correct order.

extension interface. Transport sockets follow the lifecycle events of a TCP connection and
read/write into network buffers. Some key methods that transport sockets must implement are:

.. code-block:: cpp

  virtual void onConnected() PURE;
  virtual IoResult doRead(Buffer::Instance& buffer) PURE;
  virtual IoResult doWrite(Buffer::Instance& buffer, bool end_stream) PURE;
  virtual void closeSocket(Network::ConnectionEvent event) PURE;

When data is available on a TCP connection, ``Network::ConnectionImpl::onReadReady()`` invokes the
:ref:`TLS <arch_overview_ssl>` transport socket via ``SslSocket::doRead()``. The transport socket
then performs a TLS handshake on the TCP connection. When the handshake completes,
``SslSocket::doRead()`` provides a decrypted plain text byte stream to an instance of
``Network::FilterManagerImpl``, responsible for managing the network filter chain.

.. image:: /_static/lor-transport-socket.svg
   :width: 80%
   :align: center

It’s important to note that no operation, whether it’s a TLS handshake or a pause of a filter
pipeline is truly blocking. Since Envoy is event-based, any situation in which processing requires

> additional data will lead to early event completion and yielding of the CPU to another event. When

Reviewer 3: just wanted to say I really like this paragraph. Super helpful for people coming into event-based programming for the first time.

   :align: center

Network filters are composed as a pipeline, unlike transport sockets which are one-per-connection.
Network filters come in three varieties:

* :repo:`ReadFilter <include/envoy/network/filter.h>` implementing ``onData()``, called when data is
  available from the connection (due to some request).
* :repo:`WriteFilter <include/envoy/network/filter.h>` implementing ``onWrite()``, called when data
  is about to be written to the connection (due to some response).
* :repo:`Filter <include/envoy/network/filter.h>` implementing both *ReadFilter* and *WriteFilter*.

The method signatures for the key filter methods are:

.. code-block:: cpp

  virtual FilterStatus onNewConnection() PURE;
  virtual FilterStatus onData(Buffer::Instance& data, bool end_stream) PURE;
  virtual FilterStatus onWrite(Buffer::Instance& data, bool end_stream) PURE;

As with the listener filter, the ``FilterStatus`` allows filters to pause execution of the filter
chain. For example, if a rate limiting service needs to be queried, a rate limiting network filter
would return ``Network::FilterStatus::StopIteration`` from ``onData()`` and later invoke
``continueReading()`` when the query completes.

> The last network filter for an HTTP listener is :ref:`HTTP connection manager

Reviewer 3: ```suggestion
The last network filter for a listener dealing with an HTTP request is :ref:`HTTP connection manager
```

or something similar. There's a chance someone might misinterpret HTTP listener

<arch_overview_http_conn_man>` (HCM). This is responsible for creating the HTTP/2 codec and managing
the HTTP filter chain. In our example, this is the only network filter. An example network filter
chain making use of multiple network filters would look like:

.. image:: /_static/lor-network-read.svg
   :width: 80%
   :align: center

On the response path, the network filter chain is executed in the reverse order to the request path.

.. image:: /_static/lor-network-write.svg
   :width: 80%
   :align: center

.. _life_of_a_request_http2_decoding:

5. HTTP/2 codec decoding
^^^^^^^^^^^^^^^^^^^^^^^^

> The HTTP/2 codec in Envoy is based on `nghttp2 <https://nghttp2.org/>`_. It is invoked by HCM with

Reviewer 3: ```suggestion
The HTTP/2 codec in Envoy is based on `nghttp2 <https://nghttp2.org/>`_. It is invoked by the HCM with
```

plaintext bytes from the TCP connection (after network filter chain transformation). The codec
decodes the byte stream as a series of HTTP/2 frames and demultiplexes the connection into a number
of independent HTTP streams. Stream multiplexing is a key feature in HTTP/2, providing significant
performance advantages over HTTP/1. Each HTTP stream handles a single request and response.

The codec is also responsible for handling HTTP/2 setting frames and both stream and connection

> level flow control.

Reviewer 3: is it pertinent to link to the flow control doc?

The codecs are responsible for abstracting the specifics of the HTTP connection, presenting a

> standard view to the HTTP connection manager and filter chain of a connection split into streams,

Reviewer 3: ```suggestion
standard view to the HTTP connection manager and http/l7 filter chain of a connection split into streams,
```

each with request/response headers/body/trailers. This is true regardless of whether the protocol is
HTTP/1, HTTP/2 or HTTP/3.

6. HTTP filter chain processing
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> For each HTTP stream, HCM instanties an :ref:`HTTP filter <arch_overview_http_filters>` chain,

Reviewer 3: ```suggestion
For each HTTP stream, the HCM instanties an :ref:`HTTP filter <arch_overview_http_filters>` chain,
```

.. image:: /_static/lor-http-filters.svg
   :width: 80%
   :align: center

There are three kinds of HTTP filter interfaces:

* :repo:`StreamDecoderFilter <include/envoy/http/filter.h>` with callbacks for request processing.
* :repo:`StreamEncoderFilter <include/envoy/http/filter.h>` with callbacks for response processing.
* :repo:`StreamFilter <include/envoy/http/filter.h>` implementing both `StreamDecoderFilter` and
  `StreamEncoderFilter`.

Looking at the decoder filter interface:

.. code-block:: cpp

  virtual FilterHeadersStatus decodeHeaders(RequestHeaderMap& headers, bool end_stream) PURE;
  virtual FilterDataStatus decodeData(Buffer::Instance& data, bool end_stream) PURE;
  virtual FilterTrailersStatus decodeTrailers(RequestTrailerMap& trailers) PURE;

Rather than operating on connection buffers and events, HTTP filters follow the lifecycle of an HTTP
request, e.g. ``decodeHeaders()`` takes HTTP headers as an argument rather than a byte buffer. The
returned ``FilterStatus`` provides, as with network and listener filters, the ability to manage filter
chain control flow.

When the HTTP/2 codec makes available the HTTP requests headers, these are first passed to

> ``decodeHeaders()`` in CustomFilter. If this ``FilterHeadersStatus`` is ``Continue``, HCM then

Reviewer 3: ```suggestion
``decodeHeaders()`` in CustomFilter. If the returned ``FilterHeadersStatus`` is ``Continue``, the HCM then
```

passes the headers (possibly mutated by CustomFilter) to the router filter.

Decoder and encoder-decoder filters are executed on the request path. Encoder and encoder-decoder

> filters are executed on the response path, in reverse direction. In both paths, the same filter

Reviewer 3: I think the second sentence about the same "ordering" is confusing. I see what you are saying: that the ordering is not changed, just reverse, but I think that using the diagrams below and deleting this sentence is sufficient.

chain ordering is used. Consider the following example filter chain:

.. image:: /_static/lor-http.svg
   :width: 80%
   :align: center

The request path will look like:

.. image:: /_static/lor-http-decode.svg
   :width: 80%
   :align: center

While the response path will look like:

.. image:: /_static/lor-http-encode.svg
   :width: 80%
   :align: center

When ``decodeHeaders()`` is invoked on the :ref:`router <arch_overview_http_routing>` filter, the

> route selection is finalized and a cluster is picked. HCM selects a route from its

Reviewer 3: ```suggestion
route selection is finalized and a cluster is picked. The HCM selects a route from its
```

``RouteConfiguration`` at the start of HTTP filter chain execution. This is referred to as the
*cached route*. Filters may modify headers and cause a new route to be selected, by asking HCM to

> clear the route cache. When the router filter is invoked, the route is finalized. The selected

Author: I clarified to "by asking HCM to clear the route cache and requesting HCM to reevaluate the route selection".

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The upstream endpoint connection's TLS transport socket encrypts the bytes from the HTTP/2 codec
output and writes them to a TCP socket for the upstream connection. As with :ref:`TLS transport
socket decryption <life_of_a_request_tls_decryption>`, in our example the cluster has a transport
socket configured that provides TLS transport security. The same interfaces exist for upstream and
downstream transport socket extensions.

.. image:: /_static/lor-client.svg
   :width: 70%
   :align: center

10. Response path and HTTP lifecycle
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The request, consisting of headers, and optional body and trailers, is proxied upstream, and the
response is proxied downstream. The response passes through the HTTP and network filters in the
opposite order from the request.

Various callbacks for decoder/encoder request lifecycle events will be invoked in HTTP filters, e.g.
when response trailers are being forwarded or the request body is streamed. Similarly, read/write
network filters will also have their respective callbacks invoked as data continues to flow in both
directions during a request.

> :ref:`Outlier detection <arch_overview_outlier_detection>` status for the endpoint is revised as the

Reviewer 3: I could see us deleting this sentence. Not sure if it adds much. Up to you.

request progresses.

A request completes when the upstream response reaches its end-of-stream, i.e. when trailers or the
response header/body with end-stream set are received. This is handled in
``Router::Filter::onUpstreamComplete()``.

It is possible for a request to terminate early. This may be due to (but not limited to):

* Request timeout.
* Upstream endpoint steam reset.
* HTTP filter stream reset.
* Circuit breaking.
* Unavailability of upstream resources, e.g. missing a cluster for a route.
* No healthy endpoints.
* DoS protection.
* HTTP protocol violations.

> * Local reply from either HCM or an HTTP filter. E.g. a rate limit HTTP filter returning a 429 response.

Reviewer 3: ```suggestion
* Local reply from either the HCM or an HTTP filter. E.g. a rate limit HTTP filter returning a 429 response.
```

If any of these occur, Envoy may either send an internally generated response, if upstream response
headers have not yet been sent, or will reset the stream, if response headers have already been
forwarded downstream. The Envoy :ref:`debugging FAQ <faq_overview_debug>` has further information on
interpreting these early stream terminations.

11. Post-request processing
^^^^^^^^^^^^^^^^^^^^^^^^^^^

Once a request completes, the stream is destroyed. The following also takes places:

> * :ref:`Statistics <arch_overview_statistics>` are updated (e.g. timing, active requests, upgrades,

Reviewer 3: There are some stats that have been updated prior to this. So might be worth clarifying?

  health checks). Stats are not written to the stats :ref:`sink
  <envoy_v3_api_field_config.bootstrap.v3.Bootstrap.stats_sinks>` at this point, they are batched
  and written by the main thread periodically. In our example this is a statsd sink.
* :ref:`Access logs <arch_overview_access_logs>` are written to the access log :ref:`sinks
  <arch_overview_access_logs_sinks>`. In our example this is a file access log.
* :ref:`Trace <arch_overview_tracing>` spans are finalized. If our example request was traced, a

>   trace span, describing the duration and details of the request would be created by HCM when

Reviewer 3: ```suggestion
  trace span, describing the duration and details of the request would be created by the HCM when
```

>   processing request headers and then finalized by HCM during post-request processing.

Reviewer 3: ```suggestion
  processing request headers and then finalized by the HCM during post-request processing.
```

  network. These receiving nodes will also be typically running Envoy and accepting the request via
  their ingress listeners.

.. image:: /_static/lor-topology-service-mesh.svg
   :width: 80%
   :align: center

.. image:: /_static/lor-topology-service-mesh-node.svg
   :width: 40%
   :align: center

Envoy is used in a variety of configurations beyond the service mesh. For example, it can also act
as an internal load balancer:

.. image:: /_static/lor-topology-ilb.svg
   :width: 65%
   :align: center

Or as an ingress/egress proxy on the network edge:

.. image:: /_static/lor-topology-edge.svg
   :width: 90%
   :align: center

In practice, a hybrid of these is often used, where Envoy features in a service mesh, on the edge
and as an internal load balancer. A request path may traverse multiple Envoys.

.. image:: /_static/lor-topology-hybrid.svg
   :width: 90%
   :align: center

Envoy may be configured in multi-tier topologies for scalability and reliability, with a request
first passing through an edge Envoy prior to passing through a second Envoy tier:

.. image:: /_static/lor-topology-tiered.svg
   :width: 80%
   :align: center

> In all the above cases, a request will arrive at a specific Envoy via TCP, UDP or Unix Domain

Reviewer: optional nit, Domain Sockets -> domain sockets here and below

* Transport socket, e.g. plain text, TLS, ALTS.
* Connection routing, e.g. PROXY protocol, original destination, dynamic forwarding.
* Authentication and authorization.
* Circuit breakers and outlier detection configuration and activation state.
* Many other configurations for networking, HTTP, listener, access logging, health checking, tracing
  and stats extensions.

It's helpful to focus on one at a time, so this example covers the following:

* An HTTP/2 request with :ref:`TLS <arch_overview_ssl>` over a TCP connection for both downstream
  and upstream.
* The :ref:`HTTP connection manager <arch_overview_http_conn_man>` as the only :ref:`network filter
  <arch_overview_network_filters>`.
* A hypothetical CustomFilter and the `router <arch_overview_http_routing>` filter as the :ref:`HTTP
  filter <arch_overview_http_filters>` chain.
* :ref:`Filesystem access logging <arch_overview_access_logs_sinks>`.
* :ref:`Statsd sink <envoy_v3_api_msg_config.metrics.v3.StatsSink>`.
* A single :ref:`cluster <arch_overview_cluster_manager>` with static endpoints.

We assume a static bootstrap configuration file for simplicity:

.. code-block:: yaml

  static_resources:
    listeners:

>     # We have a single listener bound to port 443.

Reviewer: optional nit 
We have a -> There is a
for consistency with the rest below

Also love the comments here.  So helpful

    - name: listener_https
      address:
        socket_address:
          protocol: TCP
          address: 0.0.0.0
          port_value: 443
      # A single listener filter exists for TLS inspector.
      listener_filters:
      - name: "envoy.filters.listener.tls_inspector"
        typed_config: {}
      # On the listener, there is a single filter chain that matches SNI for acme.com.
      filter_chains:
      - filter_chain_match:
          # This will match the SNI extracted by the TLS Inspector filter.
          server_names: ["acme.com"]
        # Downstream TLS configuration.
        transport_socket:
          name: envoy.transport_sockets.tls
          typed_config:
            "@type": type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.DownstreamTlsContext
            common_tls_context:
              tls_certificates:
              - certificate_chain: { filename: "certs/servercert.pem" }
                private_key: { filename: "certs/serverkey.pem" }
        filters:

>         # HTTP connection manager is the only network filter.

Reviewer: The HTTP connection manager ?

                routes:
                - match:
                    path: "/foo"
                  route:
                    cluster: some_service
        # CustomFilter and the HTTP router filter are the HTTP filter chain.
        http_filters:
            - name: some.customer.filter
            - name: envoy.filters.http.router
    clusters:
    - name: some_service
      connect_timeout: 5s
      # Upstream TLS configuration.
      transport_socket:
        name: envoy.transport_sockets.tls
        typed_config:
          "@type": type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.UpstreamTlsContext
      load_assignment:
        cluster_name: some_service
        # Static endpoint assignment.
        endpoints:
        - lb_endpoints:
          - endpoint:
              address:
                socket_address:
                  address: 10.1.2.10
                  port_value: 10002
          - endpoint:
              address:
                socket_address:
                  address: 10.1.2.11
                  port_value: 10002
      http2_protocol_options:
        max_concurrent_streams: 100
    - name: some_statsd_sink
      connect_timeout: 5s

>       # .. rest of configuration for statsd sink cluster

Reviewer: The rest?

``continueReading()`` when the query completes.

The last network filter for a listener dealing with HTTP is :ref:`HTTP connection manager
<arch_overview_http_conn_man>` (HCM). This is responsible for creating the HTTP/2 codec and managing
the HTTP filter chain. In our example, this is the only network filter. An example network filter
chain making use of multiple network filters would look like:

.. image:: /_static/lor-network-read.svg
   :width: 80%
   :align: center

On the response path, the network filter chain is executed in the reverse order to the request path.

.. image:: /_static/lor-network-write.svg
   :width: 80%
   :align: center

.. _life_of_a_request_http2_decoding:

5. HTTP/2 codec decoding
^^^^^^^^^^^^^^^^^^^^^^^^

The HTTP/2 codec in Envoy is based on `nghttp2 <https://nghttp2.org/>`_. It is invoked by the HCM
with plaintext bytes from the TCP connection (after network filter chain transformation). The codec
decodes the byte stream as a series of HTTP/2 frames and demultiplexes the connection into a number
of independent HTTP streams. Stream multiplexing is a key feature in HTTP/2, providing significant

> performance advantages over HTTP/1. Each HTTP stream handles a single request and response.

Reviewer: HTTP stream -> HTTP/2 stream?  or just stream?

is invoked, the route is finalized. The selected route’s configuration will point at an upstream
cluster name. The router filter then asks the `ClusterManager` for an HTTP :ref:`connection pool
<arch_overview_conn_pool>` for the cluster. This involves load balancing and the connection pool,
discussed in the next section.

.. image:: /_static/lor-route-config.svg
   :width: 70%
   :align: center

The resulting HTTP connection pool is used to build an `UpstreamRequest` object in the router, which
encapsulates the HTTP encoding and decoding callback methods for the upstream HTTP request. Once a
stream is allocated on a connection in the HTTP connection pool, the request headers are forwarded
to the upstream endpoint by the invocation of ``UpstreamRequest::encoderHeaders()``.

The router filter is responsible for all aspects of upstream request lifecycle management on the
stream allocated from the HTTP connection pool. It also is responsible for request timeouts, retries
and affinity.

7. Load balancing
^^^^^^^^^^^^^^^^^

Each cluster has a :ref:`load balancer <arch_overview_load_balancing>` which picks an endpoint when

> a new request arrives. There are a variety of load balancing algorithms that Envoy supports, e.g.

Reviewer: Envoy supports a variety of

   :align: center

> In all the above cases, a request will arrive at a specific Envoy via TCP, UDP or Unix Domain

Reviewer: ditto here on Domain Sockets -> domain sockets
